common/src/Common/Examples/Chess.hs

```diff
@@ -0,0 +1,274 @@
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE FlexibleInstances  #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes        #-}
+{-# LANGUAGE RecursiveDo       #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# LANGUAGE LambdaCase   #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE DataKinds             #-}
+
+module Common.Examples.Chess where
+-- This implements the game rules of chess, without any Reflex
+-- or frontend code.
+
+import Control.Monad
+import qualified Data.Array.IArray as A
+import Data.Foldable
+import Data.Maybe
+import Data.Ord
+import GHC.Generics
+
+type Point = (Int, Int)
+
+data Color = White | Black deriving (Eq, Show)
+data Piece = King | Queen | Rook | Bishop | Knight | Pawn deriving (Eq, Show)
+
+data ColoredPiece = ColoredPiece Color Piece deriving (Eq)
+
+instance Show ColoredPiece where
+  show (ColoredPiece clr p) = show clr <> " " <> show p
+
+type Board = A.Array Point (Maybe ColoredPiece)
+
+initialBoard :: Board
+initialBoard = A.array ((0, 0), (7,7))
+               [ ((i, j), square i j) | i <- [0..7], j <- [0..7] ]
+  where
+    initialColor n | n >= 6 = Just Black
+                   | n <= 1 = Just White
+                   | otherwise = Nothing
+    mainRow n = n == 7 || n == 0
+    pawnRow n = n == 1 || n == 6
+    piece i j | pawnRow j = Just Pawn
+              | mainRow j = case i of
+                0 -> Just Rook
+                1 -> Just Knight
+                2 -> Just Bishop
+                3 -> Just Queen
+                4 -> Just King
+                5 -> Just Bishop
+                6 -> Just Knight
+                7 -> Just Rook
+                _ -> Nothing
+              | otherwise = Nothing
+    square :: Int -> Int -> Maybe ColoredPiece
+    square i j = ColoredPiece <$> initialColor j <*> piece i j
+
+data CastleState = CastleState
+  { canCastleQueenSide :: Bool
+  , canCastleKingSide  :: Bool
+  } deriving (Show, Eq)
+
+data ChessConfig = ChessConfig
+  { configSelfCapture :: Bool
+  } deriving (Eq, Show)
+
+data GameState = GameState
+  { gameStateBoard       :: Board
+  , gameStateTurn        :: Color
+  , gameStateCastle      :: Color -> CastleState
+  , gameStatePhantomPawn :: (Maybe Point)
+  , gameStateMoveCount   :: Int
+  , gameStateConfig      :: ChessConfig
+  } deriving Eq
+
+instance Eq (Color -> CastleState) where
+  a == b = a White == b White && a Black == b Black
+
+instance Ord GameState where
+  compare = comparing gameStateMoveCount
+
+initialGS :: ChessConfig -> GameState
+initialGS config = GameState initialBoard White (const $ CastleState True True) Nothing 0 config
+
+validSquares :: [Point]
+validSquares = [(i, j) | i <- [0..7], j <- [0..7]]
+
+promotionPieces :: [Piece]
+promotionPieces = [Queen, Rook, Bishop, Knight]
+
+isValidSquare :: Point -> Bool
+isValidSquare (x, y) = x >= 0 && y >= 0 && x <= 7 && y <= 7
+
+(<!>) :: GameState -> Point -> Maybe ColoredPiece
+gs <!> point = do
```

Comment:
I think better to use name. Can use backticks to keep infix (even in definition not just usage IIRC) if you like. Don't want the people with Strong Feelings about custom infix operators to be pushed away from reflex.